packages/module-tenants/src/context/TenantAgentContextProvider.ts

```diff
@@ -0,0 +1,144 @@
+import type { AgentContextProvider, RoutingCreatedEvent, EncryptedMessage } from '@aries-framework/core'
+
+import {
+  AriesFrameworkError,
+  injectable,
+  AgentContext,
+  EventEmitter,
+  inject,
+  Logger,
+  RoutingEventTypes,
+  InjectionSymbols,
+  KeyType,
+  Key,
+  isValidJweStructure,
+  JsonEncoder,
+  isJsonObject,
+} from '@aries-framework/core'
+
+import { TenantService } from '../services'
+
+import { TenantSessionCoordinator } from './TenantSessionCoordinator'
+
+@injectable()
+export class TenantAgentContextProvider implements AgentContextProvider {
+  private tenantService: TenantService
+  private rootAgentContext: AgentContext
+  private eventEmitter: EventEmitter
+  private logger: Logger
+  private tenantSessionCoordinator: TenantSessionCoordinator
+
+  public constructor(
+    tenantService: TenantService,
+    rootAgentContext: AgentContext,
+    eventEmitter: EventEmitter,
+    tenantSessionCoordinator: TenantSessionCoordinator,
+    @inject(InjectionSymbols.Logger) logger: Logger
+  ) {
+    this.tenantService = tenantService
+    this.rootAgentContext = rootAgentContext
+    this.eventEmitter = eventEmitter
+    this.tenantSessionCoordinator = tenantSessionCoordinator
+    this.logger = logger
+
+    // Start listener for newly created routing keys, so we can register a mapping for each new key for the tenant
+    this.listenForRoutingKeyCreatedEvents()
+  }
+
+  public async getAgentContextForContextCorrelationId(tenantId: string) {
```

Comment:
What  is a 'correlation id'? Looks like this just fetches the context for a specific tenant id. `tenantId` is more clear imho.

Reply:
Yes this seems to be causing some confusion. See my reply to berend: https://github.com/hyperledger/aries-framework-javascript/pull/921#issuecomment-1177253640. The tldr is that multi-tenancy is an optional addon so I'd like to keep the word tenant/multi-tenancy out of core

Reply:
Ok, clear. Not sure I like the term `correlationId`, but I also can't think of anything better atm...

Reply:
Yes feel free to make suggestions on the name. I'm also not a fan of it, but it was the best I can come up with. It's different than a contextId (as multiple contexts could have the same tenant id).